examples/notebooks/beam-ml/alloydb_product_catalog_embeddings.ipynb

```diff
@@ -2148,6 +2160,599 @@
+        "            message[\"price\"] = round(message[\"price\"] * dynamic_factor, 2)\n",
+        "            message[\"description\"] = f\"PRICE UPDATE (factor: {dynamic_factor:.3f}): \" + message[\"description\"]\n",
+        "\n",
+        "        message[\"published_at\"] = datetime.datetime.now().isoformat()\n",
+        "\n",
+        "        data = json.dumps(message).encode('utf-8')\n",
+        "        publish_future = publisher.publish(topic_path, data)\n",
+        "\n",
+        "        try:\n",
+        "            logger.info(f\"Publishing message {message}\")\n",
+        "            file_handler.flush()\n",
+        "            message_id = publish_future.result()\n",
+        "            logger.info(f\"Published message {i+1}: {message['id']} (Message ID: {message_id})\")\n",
+        "            file_handler.flush()\n",
+        "        except Exception as e:\n",
+        "            logger.error(f\"Error publishing message: {e}\")\n",
+        "            file_handler.flush()\n",
+        "\n",
+        "        time.sleep(5)\n",
+        "\n",
+        "    logger.info(\"Finished publishing all messages.\")\n",
+        "    file_handler.flush()"
+      ],
+      "metadata": {
+        "id": "C9Bf0Nb0vY7r",
+        "cellView": "form"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "#### Start publishing to PusSub in background"
+      ],
+      "metadata": {
+        "id": "jnUSynmjEmVr"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "# Launch publisher in a separate thread\n",
+        "print(\"Starting publisher thread...\")\n",
+        "publisher_thread = threading.Thread(\n",
+        "    target=publisher_function,\n",
+        "    args=(PROJECT_ID, TOPIC),\n",
+        "    daemon=True\n",
+        ")\n",
+        "publisher_thread.start()\n",
+        "print(f\"Publisher thread started with ID: {publisher_thread.ident}\")\n",
+        "print(f\"Publisher thread logging to file: publisher_{publisher_thread.ident}.log\")"
+      ],
+      "metadata": {
+        "id": "ZnBBTwZHw7Ex"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Define Streaming Pipeline\n",
+        "Our pipeline contains these key components:\n",
+        "\n",
+        "1. **Source**: Continuously reads messages from PubSub\n",
+        "2. **Windowing**: Groups messages into 10-second windows for batch processing\n",
+        "3. **Transformation**: Converts JSON messages to Chunk objects for embedding\n",
+        "4. **ML Processing**: Generates embeddings using HuggingFace models\n",
+        "5. **Sink**: Writes results to AlloyDB with conflict resolution"
+      ],
+      "metadata": {
+        "id": "EK3pela7CLFj"
+      }
+    },
+    {
+      "cell_type": "markdown",
```

Comment:
Given that the pipeline is defined one cell down, I don't think we get much out of including it here. It also renders a little oddly in GitHub and probably in GCP docs - https://github.com/apache/beam/blob/8c588f2d5c53f637eb745478beead92e4c5e19ce/examples/notebooks/beam-ml/alloydb_product_catalog_embeddings.ipynb